File: semgus-sketch/src/main/java/org/semgus/sketch/base/Expr.java
package org.semgus.sketch.base;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

/**
 * Sketch expression.
 */
public sealed interface Expr extends Syntax {
  record Ref(@NotNull Id id) implements Expr {}

  record App(@NotNull Id fn, @NotNull List<Expr> args) implements Expr {}

  record Get(@NotNull Id id, @NotNull Id field) implements Expr {}

  record Assign(@NotNull Id id, @NotNull Expr e) implements Expr {}

  record Unary(@NotNull Id op, @NotNull Expr e) implements Expr {}

  record Binary(@NotNull Id op, @NotNull Expr l, @NotNull Expr r) implements Expr {}

  record Cond(@NotNull Expr i, @NotNull Expr t, @NotNull Expr e) implements Expr {}

  record Choice(@NotNull List<Expr> es) implements Expr {}

  record Forall(@NotNull List<Param> binds, @NotNull Expr e) implements Expr {}

  static @NotNull Expr ref(@NotNull Id id) {
    return new Ref(id);
  }

  static @NotNull Expr refPlain(@NotNull String s) {
    return ref(Id.plain(s));
  }

  static @NotNull Expr refPlain(long s) {
    return ref(Id.plain(s));
  }

  static @NotNull Expr app(@NotNull Id fn, @NotNull List<Expr> args) {
    return new App(fn, args);
  }

  static @NotNull Expr get(@NotNull Id id, @NotNull Id field) {
    return new Get(id, field);
  }

  static @NotNull Expr assign(@NotNull Id id, @NotNull Expr e) {
    return new Assign(id, e);
  }

  static @NotNull Expr assignPlain(@NotNull String idName, @NotNull Expr e) {
    return assign(Id.plain(idName), e);
  }

  static @NotNull Expr appPlain(@NotNull String fnName, @NotNull List<Expr> args) {
    return app(Id.plain(fnName), args);
  }

  static @NotNull List<Expr> bnded0(@NotNull List<Expr> args) {
    var bndExpr = Expr.refPlain("bnd");
    return Stream.concat(args.stream(), Stream.of(bndExpr)).toList();
  }

  static @NotNull List<Expr> bnded1(@NotNull List<Expr> args) {
    var bndExpr = Expr.binaryPlain("-", refPlain("bnd"), refPlain(1));
    return Stream.concat(args.stream(), Stream.of(bndExpr)).toList();
  }

  static @NotNull Expr unary(@NotNull Id op, @NotNull Expr e) {
    return new Unary(op, e);
  }

  static @NotNull Expr unaryPlain(@NotNull String opName, @NotNull Expr e) {
    return unary(Id.plain(opName), e);
  }

  static @NotNull Expr binary(@NotNull Id op, @NotNull Expr l, @NotNull Expr r) {
    return new Binary(op, l, r);
  }

  static @NotNull Expr binaryPlain(@NotNull String opName, @NotNull Expr l, @NotNull Expr r) {
    return binary(Id.plain(opName), l, r);
  }

  static @NotNull Expr cond(@NotNull Expr i, @NotNull Expr t, @NotNull Expr e) {
    return new Cond(i, t, e);
  }

  static @NotNull Expr choice(@NotNull List<Expr> es) {
    return new Choice(es);
  }

  static @NotNull Expr forall(@NotNull List<Param> binds, @NotNull Expr e) {
    return new Forall(binds, e);
  }

  static private Expr subHelper(Expr expr, Map<String, Param> oldToNew) {
    if (expr instanceof Expr.Ref && oldToNew.containsKey(((Ref) expr).id.name()))
      return Expr.ref(Id.plain(oldToNew.get(((Ref) expr).id.name()).id().name()));
    if (expr instanceof Expr.Ref || expr instanceof Expr.Get)
      return expr;
    if (expr instanceof Expr.App)
      return Expr.app(
        ((App) expr).fn,
        ((App) expr).args
          .stream()
          .map(x->subHelper(x, oldToNew))
          .toList()
      );
    if (expr instanceof Expr.Assign)
      return Expr.assign(((Assign) expr).id, subHelper(((Assign) expr).e, oldToNew));
    if (expr instanceof Expr.Unary)
      return Expr.unary(((Unary) expr).op, subHelper(((Unary) expr).e, oldToNew));
    if (expr instanceof Expr.Binary)
      return Expr.binary(((Binary) expr).op, subHelper(((Binary) expr).l, oldToNew), subHelper(((Binary) expr).r, oldToNew));
    if (expr instanceof  Expr.Cond)
      return Expr.cond(
        subHelper(((Cond) expr).i, oldToNew),
        subHelper(((Cond) expr).t, oldToNew),
        subHelper(((Cond) expr).e, oldToNew)
      );
    if (expr instanceof  Expr.Choice)
      return Expr.choice(((Choice) expr).es.stream().map(x->subHelper(x, oldToNew)).toList());
    if (expr instanceof Expr.Forall)
      return Expr.forall(((Forall) expr).binds, subHelper(((Forall) expr).e, oldToNew));
    throw new RuntimeException("This is unreachable");
  }
  static @NotNull Expr subWithNewVar(Expr e, Map<String, Param> oldToNew) {
    return subHelper(e, oldToNew);
  }
}

File: semgus-sketch/src/main/java/org/semgus/sketch/syntax/Expr.kt
package org.semgus.sketch.syntax

import org.semgus.sketch.syntax.Op.*

/**
 * Sketch expression.
 */
internal sealed class Expr : Syntax() {
  data class Ref(val id: Id) : Expr()
  data class App(val fn: Id, val args: Sequence<Expr>) : Expr()
  data class Get(val obj: Id, val field: Id) : Expr()
  data class Assign(val l: Id, val r: Expr) : Expr()
  sealed class NaryLike(open val op: Op, open val es: Sequence<Expr>) : Expr()
  data class Unary(override val op: Op, val e: Expr) : NaryLike(op, sequenceOf(e))
  data class Binary(override val op: Op, val l: Expr, val r: Expr) : NaryLike(op, sequenceOf(l, r))
  data class Nary(override val op: Op, override val es: Sequence<Expr>) : NaryLike(op, es)
  data class Ite(val i: Expr, val t: Expr, val e: Expr) : Expr()
  data class Choice(val es: Sequence<Expr>) : Expr()
  data class Forall(val binds: Sequence<Param>, val e: Expr) : Expr()

  fun toMap(outVarNames: Set<String>): Map<String, Expr> {
    val binds = mutableMapOf<String, Expr>()
    fun aux(e: Expr) {
      when (e) {
        is NaryLike -> {
          when (e.op) {
            AND -> e.es.forEach { aux(it) }
            OR -> TODO("Simplify formula in propositional logic")
            NOT -> TODO("Simplify formula in propositional logic")
            XOR -> TODO("Simplify formula in propositional logic")
            EQ -> {
              val (outExprs, nonOutExprs) = e.es.partition { it is Ref && outVarNames.contains(it.id.name) }
              outExprs
                .map { (it as Ref).id.name }
                .forEach { outIdName ->
                  nonOutExprs.forEach { nonOutExpr ->
                    if (binds.containsKey(outIdName)) TODO("Check if resolution is possible.")
                    binds[outIdName] = nonOutExpr
                  }
                }
            }
            else -> throw IllegalStateException("Cannot resolve bindings.")
          }
        }
        else -> throw IllegalStateException("Not an expression evaluated from an SMT term.")
      }
    }
    aux(this)
    return binds.toMap()
  }
}

internal fun ref(id: Id) = Expr.Ref(id)
internal fun ref(param: Param) = ref(param.id)
internal fun refPlain(s: String) = ref(idPlain(s))
internal fun refPlain(n: Long) = ref(idPlain(n))

internal fun app(fn: Id, args: Sequence<Expr>) = Expr.App(fn, args)
internal fun appPlain(fnName: String, args: Sequence<Expr>) = app(idPlain(fnName), args)

internal fun appWithField(fn: Id, args: Sequence<Expr>, fieldParams: Sequence<Param>) = app(
  fn,
  args = args
    .zip(fieldParams)
    .map { (arg, fieldParam) ->
      assign(id(fieldParam.id.name) { withField() }, arg)
    },
)

internal fun argsBnded0(args: Sequence<Expr>) = args + refPlain("bnd")
internal fun argsBnded1(args: Sequence<Expr>) = args + binary(MINUS, refPlain("bnd"), refPlain(1))

internal fun get(obj: Id, field: Id) = Expr.Get(obj, field)

internal fun getPlain(obj: Id, fieldName: String) = get(obj, id(fieldName) { withField() })

internal fun assign(l: Id, r: Expr) = Expr.Assign(l, r)
internal fun assignPlain(lName: String, r: Expr) = assign(idPlain(lName), r)

internal fun unary(op: Op, e: Expr) = Expr.Unary(op, e)
internal fun binary(op: Op, l: Expr, r: Expr) = Expr.Binary(op, l, r)
internal fun nary(op: Op, es: Sequence<Expr>) = Expr.Nary(op, es)

internal fun constraint(ntVarName: String, outputs: Sequence<Param>) = nary(
  op = AND,
  es = outputs.map { v ->
    binary(
      op = EQ,
      l = getPlain(
        obj = idPlain(ntVarName),
        fieldName = v.id.name,
      ),
      r = refPlain(v.id.name),
    )
  },
)

internal fun ite(i: Expr, t: Expr, e: Expr) = Expr.Ite(i, t, e)

internal fun choice(es: Sequence<Expr>) = Expr.Choice(es)

internal fun forall(binds: Sequence<Param>, e: Expr) = Expr.Forall(binds, e)
